README.md
# Affirm PHP API Connector

This is a stand-alone PHP library that connects to the
[Affirm](http://affirm.com) API. The intent is to be able to plug this
library into the e-commerce platform of choice, provided it can use PHP. It
can either be installed as a library to your PHP driven website or your
commandline PHP scripting library.

# Installation

Unpack or clone this git repository into a directory where you desire to place
the files. When you use this for your library, all you need is a one line
include to the `affirm-api.php` file in the root level of this directory.
This file will include all the other ones needed to load the classes required
for this API to work properly.

```php
<?php

/* Choose a directory or path to your application */
define('DOC_ROOT', '/home/website/public_html/');

/* Include the main file (which will include all others needed) */
include(DOC_ROOT . '/affirm-php/affirm-api.php');

```

## Dependencies

This library depends on [PHP](http://php.net/) 5.5 or later. Most
GNU/Linux distributions provide them in an easily downloaded and installed
repository. The [PHP cURL Library](http://php.net/manual/en/book.curl.php) and
the [PHP JSON Library](http://php.net/manual/en/book.json.php) are also needed,
and are often packaged separately. Installation instructions are shown below
for each of the major GNU/Linux distribution types. These commands must all
be run as root, or use `sudo` utility.

## APT (Debian/Ubuntu based distributions)

```shell

# Update the apt cache
apt-get update

# Install the dependencies
apt-get install php5 php5-curl php5-json
```

## YUM (Fedora/CentOS/Red Hat based distributions)

```shell

# No need to update cache, it's automatic

# Install the dependencies
yum install php5 php5-json php5-curl
```

## Affirm Account

A current [Affirm](https://www.affirm.com/) account is required for this
library. You may sign up [here](https://www.affirm.com/merchants/).
Once signed up for Affirm, a public/private key pair that serves as your
username and password would be provided, along with a Financial Product Key
and your API URL. In another area of your application that faces
the web, you should follow the steps outlined in the Affirm
[Site Integration](http://docs.affirm.com/v2/api/#site-integration)
documentation. Your server should also be set up to receive the webhook that
sends the `checkout_token` to your application as this checkout token is
used for initializing your charge (see *Usage* below).

## Configuration

One final step before starting to call objects in the application is to copy
`config.php.default` to `config.php` and update the information for your
credentials. This configuration file is used to create the `AffirmConfig` class
which is required for the application. If you choose not to enter your details
in this file because it is stored in a database, this file is still required
because the `AffirmConfig` class provides the current API base URLs.

# Usage

This library contains a few PHP classes that are used for interfacing with
Affirm and your application. You may either hard-code your details in
`config.php` or if stored in your application's database, you can enter your
credentials into the arguments of the `create_charge()` method of the
`AffirmAPI` class. The `$live_baseurl` and `$sandbox_baseurl` properties
of the `AffirmConfig` class must be left as shown in the `config.php.default`
 The methods of the `AffirmAPI` class and how to instantiate the class is shown
below. Unless you are doing testing, your code should only interface with
instances of the `AffirmAPI` class.

## Create an Instance of The AffirmAPI Class

Once the library is installed and properly linked as shown in *Installation*
above, you may create an instance of the `AffirmAPI` class in two ways. The
first way shows if you used the `AffirmConfig` class to store all of your
API keys and other credentials:

```php

// Create a new AffirmAPI with a fully configured AffirmConfig class
$affirm = new AffirmAPI();
```

The alternative is to add arguments so that it will construct the object
with these credentials overriding the ones stored in the `AffirmConfig`
class:

```php

// These would really be in a database, but showing this for example
$public_key = 'secretkey'; /**< Public API key, also used on Site Integration */
$private_key = 'supersecretkey'; /**< Private API key, never share this! */
$product_key = 'financialkey'; /**< Financial product key */
$production = true; /**< Set this to false if in sandbox mode */

// Create a new AffirmAPI with the above information included
$affirm = new AffirmAPI($public_key, $private_key, $product_key, $production);
```

## Error Handling

All methods (except the constructor) of the AffirmAPI class return a status
code while updating the object with the reponse from Affirm. The method returns
an integer status that corresponds with the Affirm status code if an error is
returned from Affirm. It returns the integer 1 if the API catches an error
before sending a request to Affirm. If a non-zero status is sent, the `response`
object in the AffirmAPI object is an Error object as defined in the 
[Affirm Documentation](http://docs.affirm.com/v2/api/errors/#status-codes). If
no error is observed, the status returned is 0 (instead of the HTTP status 200)
to behave similar to other APIs where zero mean no error.

## Create a new Charge

Your web-facing application would harvest a `checkout_token` which is required
for creating a new charge. The information for the new charge is stored in the
instance of the `AffirmAPI` class and the data can be used to do further
operations with the charge. You should also store the `charge_id` in your
application's database if you intend on updating information at any time in
the future. Once you use your `checkout_token` it is safe to discard as it is
a one-time use token and is no longer usable once a new charge is created.

```php

$token = 'exampleaffirmtoken' /**< checkout_token from Affirm */
// Creating a charge, storing data in the $affirm object
$error = $affirm->create_charge($token);
if(!$error){
  $charge_object = $affirm->response;
  $charge_id = $affirm->response->id;
}
else{
  $error_object = $affirm->response;
}

```

Once you have a `charge_id`, you can use it to perform further actions to the
charge as needed.

## Read a Charge

When you read a charge, you retreive the information that Affirm has on your
charge. The output that is registered in the instance of the `AffirmAPI` is
the same as the output from the `create_charge()` method, except this
uses the `charge_id` since you can only create a charge once.

```php

// Given a $charge_id request a charge object from Affirm
$error = $affirm->read_charge($charge_id);
if(!$error){
  $charge_object = $affirm->response;
}
else{
  $error_object = $affirm->response;
}

```

## Capture a Charge

To capture a charge, use the `capture_charge()` method with your `charge_id`
in addition to a few optional parameters. These additional parameters are
optional:

* `order_id` - A string field for your own reference
* `shipping_carrier` - A string that contains the shipping carrier
* `shipping_confirmation` - A string that contains the shipping confirmation

```php

// Given a $charge_id and no optional fields desired
$error = $affirm->capture_charge($charge_id);
if(!$error){
  $charge_event = $affirm->response;
}
else{
  $error_object = $affirm->response;
}

// If adding the optional fields, include them as follows
$order_id = 'yourownappid'; /**< Order ID for your own records */
$shipping_carrier = 'USPS'; /**< Carrier shipping the goods */
$shipping_confirmation = 'someconfnumber'; /**< Shipping confirmation number */

$error = $affirm->capture_charge($charge_id, $order_id, $shipping_carrier, $shipping_confirmation);
if(!$error){
  $charge_event = $affirm->response;
}
else{
  $error_object = $affirm->response;
}

// If you provide only a few optional fields, set others to null
$order_id = 'yourownappid';
$shipping_carrier = 'USPS';
$shipping_confirmation = null; //Shipping confirmation was not supplied

$error = $affirm->capture_charge($charge_id, $order_id, $shipping_carrier, $shipping_confirmation);
if(!$error){
  $charge_event = $affirm->response;
}
else{
  $error_object = $affirm->response;
}

```

## Void a Charge

To void a charge, use the `void_charge()` method with your `charge_id`. No
other parameters are required.

```php

// Given a $charge_id void a charge
$error = $affirm->void_charge($charge_id);
if(!$error){
  $charge_event = $affirm->response;
}
else{
  $error_object = $affirm->response;
}
```

## Refund a Charge

To refund a charge, use the `refund_charge()` method with your `charge_id` and
the amount to be refunded. This amount must either be in cents, or omitted. If
the value is omitted or `null`, the Affirm side will attempt to refund the
entire amount of the charge.

```php

// Given a $charge_id refund a charge and an amount
$amount = 399; /**< refund amount MUST be in cents */
$error = $affirm->refund_charge($charge_id, $amount);
if(!$error){
  $charge_event = $affirm->response;
}
else{
  $error_object = $affirm->response;
}

// Omit amount to refund full amount
$error = $affirm->refund_charge($charge_id);
if(!$error){
  $charge_event = $affirm->response;
}
else{
  $error_object = $affirm->response;
}
```

## Update Shipping Information

To update shipping information, use the `updates_shipping()` method with your
`charge_id` and the optional fields mentioned in *Capture a Charge*:

* `order_id` - A string field for your own reference
* `shipping_carrier` - A string that contains the shipping carrier
* `shipping_confirmation` - A string that contains the shipping confirmation

```php

// Given a $charge_id and the fields desired to be updated
$order_id = 'yourownappid'; /**< Order ID for your own records */
$shipping_carrier = 'USPS'; /**< Carrier shipping the goods */
$shipping_confirmation = 'someconfnumber'; /**< Shipping confirmation number */

$error = $affirm->update_shipping($charge_id, $order_id, $shipping_carrier, $shipping_confirmation);
if(!$error){
  $charge_event = $affirm->response;
}
else{
  $error_object = $affirm->response;
}

// If you provide only a few optional fields, set others to null
$order_id = null; //Was not changed
$shipping_carrier = 'UPS'; //Changed carrier
$shipping_confirmation = 'someconfnumber'; //Shipping confirmation changed

$error = $affirm->update_shipping($charge_id, $order_id, $shipping_carrier, $shipping_confirmation);
if(!$error){
  $charge_event = $affirm->response;
}
else{
  $error_object = $affirm->response;
}
```


CHANGELOG.txt
Affirm API 1.0.0, [2015-07-22]
- Testing suite shows success or failure for each item and at the end of tests
- Minor clarifications and bugfixes

Affirm API 0.9.0, [2015-06-22]
------------------------------
- Combined old project files to make this project
- Built up OOP framework
- Set up internal API and testing protocol
- Error and Exception Handling


test/web/index.php
<!DOCTYPE html>
<html>
  <head>
    <title>Affirm Testing</title>
  </head>
  <body>
    <pre>
<?php
$foo =  file_get_contents('php://input');
echo $foo;
?>
    </pre>
    <p><a href="tester.php">Send another test!</a></p>
  </body>
</html>


test/web/tester.php
<?php

/**
 * @file
 * Runs affirm.js to get you to fill out info for Checkout Token
 *
 * @category   Testing
 * @package    Affirm API
 * @author     Michael Sypolt <michael.sypolt@nurelm.com>
 * @copyright  Copyright (c) 2015
 * @license    http://www.gnu.org/licenses/gpl-3.0.txt
 * @version    0.5.0
 */
// Include the configuration directory
require_once ('webconfig.php');

$host = $_SERVER['HTTP_HOST'];
$root = $_SERVER['DOCUMENT_ROOT'];
$chars = strlen($root);
$dir = substr(__DIR__, $chars);
?>
<!DOCTYPE html>
<html>
  <head>
    <title>Affirm Tester</title>
    <script>
      var _affirm_config = {
      public_api_key:  "<?php echo $public_key; ?>",
      script:          "<?php echo $affirm_url; ?>"
      };
      (function(l,g,m,e,a,f,b){var d,c=l[m]||{},h=document.createElement(f),n=document.getElementsByTagName(f)[0],k=function(a,b,c){return function(){a[b]._.push([c,arguments])}};c[e]=k(c,e,"set");d=c[e];c[a]={};c[a]._=[];d._=[];c[a][b]=k(c,a,b);a=0;for(b="set add save post open empty reset on off trigger ready setProduct".split(" ");a<b.length;a++)d[b[a]]=k(c,e,b[a]);a=0;for(b=["get","token","url","items"];a<b.length;a++)d[b[a]]=function(){};h.async=!0;h.src=g[f];n.parentNode.insertBefore(h,n);delete g[f];d(g);l[m]=c})(window,_affirm_config,"affirm","checkout","ui","script","ready");
    </script>
    <script>
      // setup and configure checkout
        affirm.checkout(JSON.parse('{"config":{"financial_product_key":"<?php echo $product_key;?>"},"merchant":{"user_cancel_url":"https://<?php echo $host . $dir; ?>/","user_confirmation_url":"https://<?php echo $host . $dir; ?>/","user_confirmation_url_action":"POST"},"billing":{"name":{"first":"<?php echo $firstname; ?>","last":"<?php echo $lastname; ?>"},"address":{"line1":"<?php echo $addr; ?>","city":"<?php echo $city; ?>","state":"<?php echo $state;?>","zipcode":"<?php echo $zip; ?>","country":"<?php echo $nation; ?>"},"email":"<?php echo $email; ?>","phone_number":"<?php echo $phone; ?>","phone_number_alternative":""},"shipping":{"name":{"first":"<?php echo $firstname; ?>","last":"<?php echo $lastname; ?>"},"address":{"line1":"<?php echo $addr; ?>","city":"<?php echo $city; ?>","state":"<?php echo $state;?>","zipcode":"<?php echo $zip; ?>","country":"<?php echo $nation; ?>"},"email":"<?php echo $email; ?>","phone_number":"<?php echo $phone; ?>"},"items":[{"display_name":"Widget","sku":"w1337","unit_price":1300,"qty":1,"item_image_url":"https://<?php echo $host; ?>/img/w1337.jpg","item_url":"https://<?php echo $host;?>/product/w1337.php"}],"currency":"USD","discounts":[],"tax_amount":0,"shipping_amount":37,"total":1337}'))


        // submit and redirect to checkout flow
        affirm.checkout.post();
    </script>
  </head>
  <body>

  </body>
</html>


test/affirm-test.php
<?php

/**
 * @file
 * Tester file for Affirm API Connector
 *
 * @category   Testing
 * @package    Affirm API
 * @author     Michael Sypolt <michael.sypolt@nurelm.com>
 * @copyright  Copyright (c) 2015
 * @license    http://www.gnu.org/licenses/gpl-3.0.txt
 * @version    0.5.0
 */
//Include the affirm includer
define('TEST_ROOT', __DIR__);
require_once(TEST_ROOT . '/../affirm_api.php');

$error = 0;     /**< Try to make some errors if set to 1 */
$help = 0;      /**< Show help screen if set to 1 */
$long = 0;      /**< Build object in long form if set to 1 */
$nocapture = 0; /**< Do not capture charge, test voiding */
$ship = 0;      /**< If set to 1, use shipping info tests for updating */
$token = '';    /**< Token to be sent to Affirm in testing, only once */
$verbose = 0;   /**< If set to 1, shows var_dump() of data from Affirm */

$failures = 0;  /**< Number of failures observed in this test */

if (count($_SERVER['argv']) > 10){
  echo "Don't be so argumentative!\n";
  $help = 1;
}
else{
  foreach($_SERVER['argv'] as $i => $argument){
    if($i > 0){
      $subargs = explode('=', $argument);
      // Find the optional flags
      if (strpos('error', $subargs[0]) === 0){$error = 1;}
      if (strpos('help', $subargs[0]) === 0){$help = 1;}
      if (strpos('long', $subargs[0]) === 0){$long = 1;}
      if (strpos('nocapture', $subargs[0]) === 0){$nocapture = 1;}
      if (strpos('shipping', $subargs[0]) === 0){$ship = 1;}
      if (strpos('verbose', $subargs[0]) === 0){$verbose = 1;}

      // Find the token
      if (strpos('token', $subargs[0]) === 0 && count($subargs)>1){
        $token = $subargs[1];
      }
    }
  }
}
if ($token == '' || $help == 1){
  echo "Usage: php affirm_api.php [option(s)]\n";
  echo "Test the affirm library\n\n";
  echo "The following are the available options\n\n";
  echo "  error         Coerce as many errors as possible, for further testing\n";
  echo "  help          Shows this help screen and exits\n";
  echo "  long          Build the object with arguments instead of using AffirmConfig\n";
  echo "  nocapture     Do not capture to test voiding a charge\n";
  echo "  shipping      Use the additional optional shipping info for capture/update\n";
  echo "  token=TOKEN   REQUIRED: checkout_token from Affirm \n";
  echo "  verbose       Verbose option to var_dump the objects received\n\n";
  echo "Any of the first few letters can be used as a shortcut for all options\n";
  exit(1);
}

if($long == 1){
  // Create a new AffirmAPI with the above information included
  $config = new AffirmConfig();

  $public_key = $config->sandbox_pubic_key;
  $private_key = $config->sandbox_private_key;
  $product_key = $config->product_key;
  $production = false; /**< Set this to false if in sandbox mode */

  $affirm = new AffirmAPI($public_key, $private_key, $product_key, $production);
}
else{
  // Create a new AffirmAPI with a fully configured AffirmConfig class
  $affirm = new AffirmAPI();
}

// Creating a charge, storing data in the $affirm object
echo "Creating Charge from token {$token}: ";
$status = $affirm->create_charge($token);
if ($status != 0){
  $failures += 1;
  $success = 'FAIL';
}
else{
  $success = 'SUCCESS';
}
echo "{$status} {$success}\n";
if ($verbose == 1){
  var_dump($affirm->response);
}
if ($status == 0){
  $charge_id = $affirm->response->id;
}
else{
  echo "Token was already used or invalid, no point in further testing :(\n";
  echo "Please use your site integration (contents of web directory) to make a token\n";
  exit(2);
}

if($error == 1){
  echo "Trying to make Affirm send an error by reusing the token: ";
  $status = $affirm->create_charge($token);
  if ($status != 400){
    $failures += 1;
    $success = 'FAIL';
  }
  else{
    $success = 'SUCCESS';
  }
  echo "{$status} {$success}\n";
  if ($verbose == 1){
    var_dump($affirm->response);
  }

  echo "Trying to make Affirm send an error by using a silly string: ";
  $status = $affirm->create_charge("somesillyinvalidstring");
  if ($status != 400){
    $failures += 1;
    $success = 'FAIL';
  }
  else{
    $success = 'SUCCESS';
  }
  echo "{$status} {$success}\n";
  if ($verbose == 1){
    var_dump($affirm->response);
  }

}

// Given a $charge_id request a charge object from Affirm
echo "Now reading a Charge: ";
$status = $affirm->read_charge($charge_id);
if ($status != 0){
  $failures += 1;
  $success = 'FAIL';
}
else{
  $success = 'SUCCESS';
}
echo "{$status} {$success}\n";
if($verbose == 1){
  var_dump($affirm->response);
}

if ($error == 1){
  echo "Trying to refund $3.99, before capturing to generate an error: ";
  // Given a $charge_id refund a charge
  $refund = 3.99; /**< refund amount is in dollars */
  $status = $affirm->refund_charge($charge_id, $refund);
  if ($status != 400){
    $failures += 1;
    $success = 'FAIL';
  }
  else{
    $success = 'SUCCESS';
  }
  echo "{$status} {$success}\n";
  if ($verbose == 1){
    var_dump($affirm->response);
  }
}

if ($nocapture == 1){
  echo "Voiding a Charge: ";
  // Given a $charge_id void a charge
  $status = $affirm->void_charge($charge_id);
  if ($status != 0){
    $failures += 1;
    $success = 'FAIL';
  }
  else{
    $success = 'SUCCESS';
  }
  echo "{$status} {$success}\n";
  if ($verbose == 1){
    var_dump($affirm->response);
  }
}

echo "Capturing a charge ";
if ($ship == 1){
  echo "with shipping info fields: ";
  // If adding the optional fields, include them as follows
  $order_id = 'yourownappid'; /**< Order ID for your own records */
  $shipping_carrier = 'UPS'; /**< Carrier shipping the goods */
  $shipping_confirmation = 'someconfnumber'; /**< Shipping confirmation number */

  $status = $affirm->capture_charge($charge_id, $order_id, $shipping_carrier, $shipping_confirmation);
}
else{
  echo ": ";
  // Given a $charge_id and no optional fields desired
  $status = $affirm->capture_charge($charge_id);
}
if ($status != 0){
  $failures += 1;
  $success = 'FAIL';
}
else{
  $success = 'SUCCESS';
}
echo "{$status} {$success}\n";
if ($verbose == 1){
  var_dump($affirm->response);
}

echo "Trying to refund $3.99, after capturing: ";
$refund = 399; 
$status = $affirm->refund_charge($charge_id, $refund);
if ($status != 0){
  $failures += 1;
  $success = 'FAIL';
}
else{
  $success = 'SUCCESS';
}
echo "{$status} {$success}\n";
if ($verbose == 1){
  var_dump($affirm->response);
}

if ($error == 1){
  echo "Trying to voiding a charge after capturing: ";
  // Given a $charge_id void a charge
  $status = $affirm->void_charge($charge_id);
  if ($status != 400){
    $failures += 1;
    $success = 'FAIL';
  }
  else{
    $success = 'SUCCESS';
  }
  echo "{$status} {$success}\n";
  if ($verbose == 1){
    var_dump($affirm->response);
  }
}
if ($ship == 1){
  echo "Since we have some shipping Info, let's try to update it: ";
  // Given a $charge_id and the fields desired to be updated
  $order_id = 'yourownappid'; /**< Order ID for your own records */
  $shipping_carrier = 'USPS'; /**< Carrier shipping the goods */
  $shipping_confirmation = 'somenewconf'; /**< Shipping confirmation number */

  $status = $affirm->update_shipping($charge_id, $order_id, $shipping_carrier, $shipping_confirmation);
  if ($status != 0){
    $failures += 1;
    $success = 'FAIL';
  }
  else{
    $success = 'SUCCESS';
  }
  echo "{$status} {$success}\n";
  if ($verbose == 1){
    var_dump($affirm->response);
  }

  echo "Try updating with a limited set of optional fields: ";
  // If you provide only a few optional fields, set others to null
  $order_id = null; //Was not changed
  $shipping_carrier = 'UPS'; //Changed carrier
  $shipping_confirmation = 'someconfnumber'; //Shipping confirmation changed

  $status = $affirm->update_shipping($charge_id, $order_id, $shipping_carrier, $shipping_confirmation);
  if ($status != 0){
    $failures += 1;
    $success = 'FAIL';
  }
  else{
    $success = 'SUCCESS';
  }
  echo "{$status} {$success}\n";
  if ($verbose == 1){
    var_dump($affirm->response);
  }
}

echo "Trying to refund balance by not passing an amount: ";
$status = $affirm->refund_charge($charge_id);
if ($status != 0){
  $failures += 1;
  $success = 'FAIL';
}
else{
  $success = 'SUCCESS';
}
echo "{$status} {$success}\n";
if ($verbose == 1){
  var_dump($affirm->response);
}

if ($failures == 0){
  $success = 'SUCCESS';
}
else{
  $success = 'FAIL';
}
echo "{$success}! Test complete with {$failures} errors.\n";


affirm_api.php
<?php

/**
 * @file
 * Bootstrap file for Affirm API Connector
 *
 * This includes all the necessary
 * to access the router, which will then access appropriate controllers
 *
 * @category   Bootstrap
 * @package    Affirm API
 * @author     Michael Sypolt <michael.sypolt@nurelm.com>
 * @copyright  Copyright (c) 2015
 * @license    http://www.gnu.org/licenses/gpl-3.0.txt
 * @version    0.9.0
 */

// Make sure this directory is application root
$AFFIRM_ROOT = __DIR__;

// Load the settings
require_once ($AFFIRM_ROOT . '/config.php');

// Load the classes
$PATH = $AFFIRM_ROOT . '/classes';
$includes = scandir($PATH);
foreach ($includes as $include){
  if (is_file($PATH . '/' . $include) && $include != '.' && $include != '..' && fnmatch("*.php", $include)){
    require_once ($PATH . '/' . $include);
  }
}


classes/api.php
<?php

/**
 * @file
 * AffirmAPI Class
 *
 * Handles user input and provides output
 *
 * @category Request Handling
 * @package Affirm API
 * @author Michael Sypolt <michael.sypolt@nurelm.com>
 * @copyright Copyright (c) 2015
 * @license http://www.gnu.org/licenses/gpl-3.0.txt
 * @version Release: @package_version@
 *
 */
class AffirmAPI {
  public $private_key;    /**< Affirm Private Key */
  public $public_key;     /**< Affirm Public Key */
  public $product_key;   /**< Affirm Financial Product Key */
  public $base_url;       /**< Affirm base url for making request */
  public $curl;           /**< AffirmCurl object */

  public $response;       /**< Data from Affirm as an object */
  public $status;         /**< HTTP Status from Affirm */

  /**
   * Constructor, uses default configuration file if left empty
   *
   * @param boolean $production Set this to false for sandbox mode
   * @param string $public_key Public Key for Affirm API
   * @param string $private_key Private Key for Affirm API
   * @param string $product_key Finanical product keyfor Affirm
   */
  public function __construct($production = null, $public_key = null, $private_key = null, $product_key = null){
    // Set up configuration, use override for production mode if needed
    if (!class_exists('AffirmConfig')){
      throw new Exception('AffirmConfig class does not exist');
    }
    $config = new AffirmConfig();
    if (is_bool($production)){
      $config->production = $production;
    }
    if ($config->production == true){
      $this->public_key = $config->live_public_key;
      $this->private_key = $config->live_private_key;
      $this->base_url = $config->live_baseurl;
    }
    else{
      $this->public_key = $config->sandbox_public_key;
      $this->private_key = $config->sandbox_private_key;
      $this->base_url = $config->sandbox_baseurl;
    }
    $this->product_key = $config->product_key;

    // Add any overrides to keys and codes
    if (is_string($public_key) && $public_key != ''){
      $this->public_key = $public_key;
    }
    if (is_string($private_key) && $private_key != ''){
      $this->private_key = $private_key;
    }
    if (is_string($product_key) && $product_key != ''){
      $this->product_key = $product_key;
    }
  }

  /**
   * Creates a new charge from the checkout_token
   *
   * @param string $checkout_token Authentication token from affirm.js
   *
   * @return int $status Status as an integer, 0 meaning success
   */
  public function create_charge($checkout_token){
    if (is_null($checkout_token) || $checkout_token == ''){
      // Make something look like the Affirm error object
      $this->response = (object) array(
        'status_code' => 1,
        'type' => 'invalid_request',
        'code' => 'invalid_field',
        'message' => 'Checkout Token is empty, could not send to Affirm',
        'field' => 'checkout_token',
      );
      return 1;
    }
    else{
      $auth_array = array('checkout_token' => $checkout_token);
      $this->curl = new AffirmCurl("https://{$this->public_key}:{$this->private_key}@{$this->base_url}", 'POST', $auth_array);
      $this->curl->send();
      $this->curl->unpack();
      $this->response = $this->curl->response_object;
      $this->status = $this->curl->status;
      if ($this->curl->status == 200){
        // Zero is always a good number to return when there is no error
        return 0;
      }
      else{
        // Just pass the status code from Affirm through the API
        return $this->curl->status;
      }
    }
  }

  /**
   * Reads a charge from a previously created charge
   *
   * @param string $charge_id Unique Charge ID from Affirm
   *
   * @return int $status Status as an integer, 0 meaning success
   */
  public function read_charge($charge_id){
    if ($charge_id == ''){
      // Make something look like the Affirm error object
      $this->response = (object) array(
        'status_code' => 1,
        'type' => 'invalid_request',
        'code' => 'invalid_field',
        'message' => 'Charge ID is empty, could not send to Affirm',
        'field' => 'charge_id',
      );
      return 1;
    }
    else{
      $this->curl = new AffirmCurl("https://{$this->public_key}:{$this->private_key}@{$this->base_url}/{$charge_id}");
      $this->curl->send();
      $this->curl->unpack();
      $this->response = $this->curl->response_object;
      $this->status = $this->curl->status;
      if ($this->curl->status == 200){
        // Zero is always a good number to return when there is no error
        return 0;
      }
      else{
        // Just pass the status code from Affirm through the API
        return $this->curl->status;
      }
    }
  }

  /**
   * Captures the charge
   *
   * @param string $charge_id Unique Charge ID from Affirm
   * @param string $order_id Optional order ID for your records
   * @param string $carrier Optional shipping carrier name
   * @param string $confirmation Optional confirmation number
   *
   * @return int $status Status as an integer, 0 meaning success
   */
  public function capture_charge($charge_id, $order_id = null, $carrier = null, $confirmation = null){
    if ($charge_id == ''){
      // Make something look like the Affirm error object
      $this->response = (object) array(
        'status_code' => 1,
        'type' => 'invalid_request',
        'code' => 'invalid_field',
        'message' => 'Charge ID is empty, could not send to Affirm',
        'field' => 'charge_id',
      );
      return 1;
    }
    else{
      $inputs = array();
      if (!is_null($order_id)){
        $inputs['order_id'] = "{$order_id}";
      }
      if (!is_null($carrier)){
        $inputs['shipping_carrier'] = "{$carrier}";
      }
      if (!is_null($order_id)){
        $inputs['shipping_confirmation'] = "{$confirmation}";
      }
      if(count($inputs) == 0){
        $inputs = '{}';
      }
      $this->curl = new AffirmCurl("https://{$this->public_key}:{$this->private_key}@{$this->base_url}/{$charge_id}/capture", 'POST', $inputs);
      $this->curl->send();
      $this->curl->unpack();
      $this->response = $this->curl->response_object;
      $this->status = $this->curl->status;
      if ($this->curl->status == 200){
        // Zero is always a good number to return when there is no error
        return 0;
      }
      else{
        // Just pass the status code from Affirm through the API
        return $this->curl->status;
      }
    }
  }

  /**
   * Voids the charge
   *
   * @param string $charge_id Unique Charge ID from Affirm
   *
   * @return int $status Status as an integer, 0 meaning success
   */
  public function void_charge($charge_id){
    if ($charge_id == ''){
      // Make something look like the Affirm error object
      $this->response = (object) array(
        'status_code' => 1,
        'type' => 'invalid_request',
        'code' => 'invalid_field',
        'message' => 'Charge ID is empty, could not send to Affirm',
        'field' => 'charge_id',
      );
      return 1;
    }
    else{
      $this->curl = new AffirmCurl("https://{$this->public_key}:{$this->private_key}@{$this->base_url}/{$charge_id}/void", 'POST', '{}');
      $this->curl->send();
      $this->curl->unpack();
      $this->response = $this->curl->response_object;
      $this->status = $this->curl->status;
      if ($this->curl->status == 200){
        // Zero is always a good number to return when there is no error
        return 0;
      }
      else{
        // Just pass the status code from Affirm through the API
        return $this->curl->status;
      }
    }
  }

  /**
   * Refunds the charge
   *
   * @param string $charge_id Unique Charge ID from Affirm
   * @param int $amount Amount of refund in cents
   *
   * @return int $status Status as an integer, 0 meaning success
   */
  public function refund_charge($charge_id, $amount=null){
    if ($charge_id == ''){
      // Make something look like the Affirm error object
      $this->response = (object) array(
        'status_code' => 1,
        'type' => 'invalid_request',
        'code' => 'invalid_field',
        'message' => 'Charge ID is empty, could not send to Affirm',
        'field' => 'charge_id',
      );
      return 1;
    }
    else{
      $inputs = array('amount' => $amount);
      $this->curl = new AffirmCurl("https://{$this->public_key}:{$this->private_key}@{$this->base_url}/{$charge_id}/refund", 'POST', $inputs);
      $this->curl->send();
      $this->curl->unpack();
      $this->response = $this->curl->response_object;
      $this->status = $this->curl->status;
      if ($this->curl->status == 200){
        // Zero is always a good number to return when there is no error
        return 0;
      }
      else{
        // Just pass the status code from Affirm through the API
        return $this->curl->status;
      }
    }
  }

  /**
   * Updates the shipping information
   *
   * @param string $charge_id Unique Charge ID from Affirm
   * @param string $order_id Optional order ID for your records
   * @param string $carrier Optional shipping carrier name
   * @param string $confirmation Optional confirmation number
   *
   * @return int $status Status as an integer, 0 meaning success
   */
  public function update_shipping($charge_id, $order_id = null, $carrier = null, $confirmation = null){
    if ($charge_id == ''){
      // Make something look like the Affirm error object
      $this->response = (object) array(
        'status_code' => 1,
        'type' => 'invalid_request',
        'code' => 'invalid_field',
        'message' => 'Charge ID is empty, could not send to Affirm',
        'field' => 'charge_id',
      );
      return 1;
    }
    else{
      $inputs = array();
      if (!is_null($order_id)){
        $inputs['order_id'] = "{$order_id}";
      }
      if (!is_null($carrier)){
        $inputs['shipping_carrier'] = "{$carrier}";
      }
      if (!is_null($order_id)){
        $inputs['shipping_confirmation'] = "{$confirmation}";
      }
      if(count($inputs) == 0){
        $inputs = '{}';
      }
      $this->curl = new AffirmCurl("https://{$this->public_key}:{$this->private_key}@{$this->base_url}/{$charge_id}/update", 'POST', $inputs);
      $this->curl->send();
      $this->curl->unpack();
      $this->response = $this->curl->response_object;
      $this->status = $this->curl->status;
      if ($this->curl->status == 200){
        // Zero is always a good number to return when there is no error
        return 0;
      }
      else{
        // Just pass the status code from Affirm through the API
        return $this->curl->status;
      }
    }
  }
}
